runtime/unix.c

```diff
@@ -431,31 +431,22 @@ char *caml_secure_getenv (char const *var)
 #endif
 }
 
-int64_t caml_time_counter(void)
+uint64_t caml_time_counter(void)
 {
-#if defined(HAS_MACH_ABSOLUTE_TIME)
-  static mach_timebase_info_data_t time_base = {0};
-  uint64_t now;
-
-  if (time_base.denom == 0) {
-    if (mach_timebase_info(&time_base) != KERN_SUCCESS)
-      return 0;
-  }
-
-  now = mach_absolute_time();
-  return (int64_t)((now * time_base.numer) / time_base.denom);
+#if defined(HAS_CLOCK_GETTIME_NSEC_NP)
+  return (clock_gettime_nsec_np(CLOCK_UPTIME_RAW));
 #elif defined(HAS_POSIX_MONOTONIC_CLOCK)
   struct timespec t;
   clock_gettime(CLOCK_MONOTONIC, &t);
   return
-    (int64_t)t.tv_sec  * (int64_t)1000000000 +
-    (int64_t)t.tv_nsec;
+    t.tv_sec  * 1000000000 +
```

Comment:
Forgot the comment.

I would have switched the coercions to `uint64_t` rather than drop them. But I'm especially bad at the promotion game :-)

Reply:
I was wondering about that too... I think that the implicit promotions will do the right thing here since uint64 is the highest rank C type here, but my head starts to hurt after reading all the implicit promotion rules. Explicit casts are probably safest here.

Reply:
> The implicit promotions will do the right thing here

No.  `t.tv_sec` has type `time_t`, which can still be 32 bits on old systems, and in this case the product `t.tv_sec * 1000000000` will be performed with a 32-bit multiplication, and overflow most of the time.

Please convert every operand of the formula to the desired result type `uint64_t` so that there is obviously no risk of overflow.

Reply:
I forgot about Y2038! Now explicit casts.